### PR TITLE
*: add pushdown for ShowStmt and implement for show columns

### LINKED ISSUE #29910: `show full columns like` should be case insensitive
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
mysql> create table t(id int);
Query OK, 0 rows affected (0.01 sec)

mysql> show full columns from t like 'id';
+-------+---------+-----------+------+------+---------+-------+---------------------------------+---------+
| Field | Type    | Collation | Null | Key  | Default | Extra | Privileges                      | Comment |
+-------+---------+-----------+------+------+---------+-------+---------------------------------+---------+
| id    | int(11) | NULL      | YES  |      | NULL    |       | select,insert,update,references |         |
+-------+---------+-----------+------+------+---------+-------+---------------------------------+---------+
1 row in set (0.00 sec)

mysql> show full columns from t like 'ID';
Empty set (0.00 sec)
```

### 2. What did you expect to see? (Required)

In MySQL 8.0:
```sql
mysql> show full columns from t like 'ID';
+-------+------+-----------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+------+-----------+------+-----+---------+-------+---------------------------------+---------+
| id    | int  | NULL      | YES  |     | NULL    |       | select,insert,update,references |         |
+-------+------+-----------+------+-----+---------+-------+---------------------------------+---------+
1 row in set (0.05 sec)
```

### 3. What did you see instead (Required)

```sql
mysql> show full columns from t like 'ID';
Empty set (0.00 sec)
```

This leads to incompatibility with some ecosystem tools.

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #29910

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -2871,6 +2871,15 @@ func (b *PlanBuilder) buildShow(ctx context.Context, show *ast.ShowStmt) (Plan,
 	}.Init(b.ctx)
 	isView := false
 	isSequence := false
+	switch show.Tp {
+	case ast.ShowColumns:
```
Comment (djshow832) [inline planner/core/planbuilder.go:2875]: Why not merge this switch-case to the switch-case below?

File: planner/core/memtable_predicate_extractor_test.go [hunk 2]
```diff
@@ -1659,11 +1661,18 @@ func TestPredicateQuery(t *testing.T) {
...
 	tk.MustQuery("select column_name from information_schema.columns where table_schema = 'TEST' and (column_name like 'I%' or column_name like '%D')").Check(testkit.Rows("id"))
 	tk.MustQuery("select column_name from information_schema.columns where table_schema = 'TEST' and (column_name like 'abc%' and column_name like '%lmn')").Check(testkit.Rows("abclmn"))
+	tk.MustQuery("describe t").Check(testkit.Rows("id int(11) YES  <nil> ", "abclmn int(11) YES  <nil> "))
+	tk.MustQuery("describe t id").Check(testkit.Rows("id int(11) YES  <nil> "))
+	tk.MustGetErrCode("describe t 'I%'", errno.ErrParse)
+	tk.MustGetErrCode("describe t I%", errno.ErrParse)
+
+	tk.MustQuery("show columns from t like `ABCLMN`").Check(testutil.RowsWithSep(",", "abclmn,int(11),YES,,<nil>,"))
+	tk.MustQuery("show columns from t like 'abc%'").Check(testutil.RowsWithSep(",", "abclmn,int(11),YES,,<nil>,"))
+	tk.MustQuery("show columns from t like '%lmn'").Check(testutil.RowsWithSep(",", "abclmn,int(11),YES,,<nil>,"))
```
Comment (djshow832) [inline planner/core/memtable_predicate_extractor_test.go:1677]: Please have more test cases according to https://docs.pingcap.com/zh/tidb/stable/sql-statement-show-columns-from/, such as
- show columns from t where field='id';
- show fields in t where field like 'abc%';

Comment (hawkingrei) [reply planner/core/memtable_predicate_extractor_test.go:1677]: Now, ``` show XXXX like XXXX``` has been implemented. So ``` where``` is not supported. I will add the ```rewrite``` to implement it for ```where```.

File: planner/core/planbuilder.go (hunk 1 above)
Comment (hawkingrei) [reply planner/core/planbuilder.go:2876]: Done

File: planner/core/show_predicate_extractor.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func (e *ShowColumnsTableExtractor) explainInfo() string {
+	r := new(bytes.Buffer)
+	if len(e.Field) > 0 {
+		r.WriteString(fmt.Sprintf("feild:[%s], ", e.Field))
+	}
+
+	if len(e.FieldPatterns) > 0 {
+		r.WriteString(fmt.Sprintf("feild_pattern:[%s], ", e.FieldPatterns))
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply planner/core/show_predicate_extractor.go:88]: Fix include testing.

File: planner/core/show_predicate_extractor.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		return true
+	}
+	return false
+
+}
+
+func (e *ShowColumnsTableExtractor) explainInfo() string {
+	r := new(bytes.Buffer)
+	if len(e.Field) > 0 {
+		r.WriteString(fmt.Sprintf("feild:[%s], ", e.Field))
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply planner/core/show_predicate_extractor.go]: Fix include testing.

**Comment by hawkingrei** [comment]
@djshow832 @wjhuang2016 PTAL

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/memtable_predicate_extractor_test.go (hunk 2 above)
Comment (djshow832) [reply planner/core/memtable_predicate_extractor_test.go:1677]: `where` is already supported.

File: planner/core/show_predicate_extractor.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+// and push the predicates down to the data retrieving on reading memory table stage when use ShowStmt.
+//
+// e.g:
+// SHOW COLUMNS FROM t LIKE '%abc%'
+// We must request all components from the memory table, and filter the result by the PatternLikeExpr predicate.
+//
+// it is a way to fix https://github.com/pingcap/tidb/issues/29910.
+type ShowPredicateExtractor interface {
+	// Extracts predicates which can be pushed down and returns the remained predicates
+	Extract(show *ast.ShowStmt) bool
```
Comment (xiongjiwei) [inline planner/core/show_predicate_extractor.go:40]: comments need update

File: planner/core/memtable_predicate_extractor_test.go (hunk 2 above)
Comment (hawkingrei) [reply planner/core/memtable_predicate_extractor_test.go:1677]: > `where`

You say the query is from ```information_schema.columns``` . but the implementation of ShowStmt  is difference from ```memtable``` in the ```information_schema.columns```. I have tested the ``` where```. it is not case-insensitivity

Comment (djshow832) [reply planner/core/memtable_predicate_extractor_test.go:1677]: I mean `show columns from t1 where field='s1';`

Comment (hawkingrei) [reply planner/core/memtable_predicate_extractor_test.go:1677]: OK, I can add it.

Comment (hawkingrei) [reply planner/core/memtable_predicate_extractor_test.go:1677]: Done

File: planner/core/show_predicate_extractor.go (hunk 5 above)
Comment (hawkingrei) [reply planner/core/show_predicate_extractor.go]: Updated

**Review by djshow832 — APPROVED** [review]

**Pull request merged**